cloudinit/osys/windows/network.py

```diff
@@ -0,0 +1,415 @@
+        if not p:
+            raise Exception(
+                'Unable to allocate memory for the IP forward table')
+        p_forward_table = ctypes.cast(
+            p, ctypes.POINTER(iphlpapi.Win32_MIB_IPFORWARDTABLE))
+
+        try:
+            err = iphlpapi.GetIpForwardTable(p_forward_table,
+                                             ctypes.byref(size), 0)
+            if err == iphlpapi.ERROR_INSUFFICIENT_BUFFER:
+                kernel32.HeapFree(heap, 0, p_forward_table)
+                p = kernel32.HeapAlloc(heap, 0, ctypes.c_size_t(size.value))
+                if not p:
+                    raise Exception(
+                        'Unable to allocate memory for the IP forward table')
+                p_forward_table = ctypes.cast(
+                    p, ctypes.POINTER(iphlpapi.Win32_MIB_IPFORWARDTABLE))
+
+            err = iphlpapi.GetIpForwardTable(p_forward_table,
+                                             ctypes.byref(size), 0)
+            if err != kernel32.ERROR_NO_DATA:
+                if err:
+                    raise Exception(
+                        'Unable to get IP forward table. Error: %s' % err)
+
+                forward_table = p_forward_table.contents
+                table = ctypes.cast(
+                    ctypes.addressof(forward_table.table),
+                    ctypes.POINTER(iphlpapi.Win32_MIB_IPFORWARDROW *
+                                   forward_table.dwNumEntries)).contents
+
+                i = 0
+                while i < forward_table.dwNumEntries:
+                    row = table[i]
+                    destination = ws2_32.Ws2_32.inet_ntoa(
+                        row.dwForwardDest).decode()
+                    netmask = ws2_32.Ws2_32.inet_ntoa(
+                        row.dwForwardMask).decode()
+                    gateway = ws2_32.Ws2_32.inet_ntoa(
+                        row.dwForwardNextHop).decode()
+                    index = ws2_32.Ws2_32.dwForwardIfIndex
+                    flags = ws2_32.Ws2_32.dwForwardProto
+                    metric = row.dwForwardMetric1
+                    route = Route(destination=destination,
+                                  gateway=gateway,
+                                  netmask=netmask,
+                                  interface=index,
+                                  metric=metric,
+                                  flags=flags)
+                    routing_table.append(route)
+                    i += 1
+
+            return routing_table
+        finally:
+            kernel32.HeapFree(heap, 0, p_forward_table)
+
+    @classmethod
+    def add(cls, route):
+        """Add a new route in the underlying OS.
+
+        The function should expect an instance of :class:`BaseRoute`.
+        """
+        args = ['ROUTE', 'ADD',
+                route.destination,
+                'MASK', route.netmask, route.gateway]
+        popen = subprocess.Popen(args, shell=False,
```

Comment:
We should likely take over the code from http://bazaar.launchpad.net/~cloud-init-dev/cloud-init/trunk/view/head:/cloudinit/util.py#L1627 (the ``subp`` function there) and use that instead, since its better and more generic for the use-cases here. Put that in cloudinit/util.py and use it?